assembly/src/assembler/dead_code_elimination.rs

```diff
@@ -0,0 +1,155 @@
+use alloc::{
+    collections::{BTreeMap, BTreeSet},
+    vec::Vec,
+};
+
+use vm_core::mast::{MastForest, MastNode, MastNodeId};
+
+/// Returns a `MastForest` where all nodes that are unreachable from all procedures are removed.
+///
+/// It also returns the map from old node IDs to new node IDs; or `None` if the `MastForest` was
+/// unchanged. Any [`MastNodeId`] used in reference to the old [`MastForest`] should be remapped
+/// using this map.
+pub fn dead_code_elimination(
+    mast_forest: MastForest,
+) -> (MastForest, Option<BTreeMap<MastNodeId, MastNodeId>>) {
```

Comment:
nit: I would probably name this: `prune_unreachable_nodes()` or something like that (i.e., would use an action verb).

Also, should we move this into `MastForest`? It could look something like:
```rust
impl MastForest {
    pub fn prune_unreachable_nodes(&mut self) -> Option<BTreeMap<MastNodeId, MastNodeId>> {
        ...
    }
}
```
The internal logic could remain the same - i.e., we could do something like this at the end of the function:
```rust
self.nodes = new_nodes;
self.roots = new_roots;
```

Reply:
Moved the code to `MastForest` (which required a bit more massaging)